<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Load FastqQC reports and other NGS related files â€¢ ngsReports</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Load FastqQC reports and other NGS related files">
<meta property="og:description" content="This package provides methods and object classes for parsing 
    FastQC reports and output summaries from other NGS tools into R.
    As well as parsing files, multiple plotting methods have been implemented 
    for visualising the parsed data.
    Plots can be generated as static ggplot objects or interactive plotly 
    objects.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">ngsReports</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.5.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/ngsReportsIntroduction.html">An Introduction to ngsReports</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/smped/ngsReports" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div class="section level1">
<div class="page-header"><h1 id="ngsreports-">ngsReports <img id="ngsreports_logo" src="reference/figures/ngsReports.png" align="right" width="125"><a class="anchor" aria-label="anchor" href="#ngsreports-"></a>
</h1></div>
<!-- badges: start -->

<p>An R Package for managing FastQC reports and other NGS related log files inside R. Except for some periodic minor bug fixes, this branch is the current release which is also available from <a href="https://bioconductor.org/packages/release/bioc/html/ngsReports.html" class="external-link">the Bioconductor website</a>.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install required packages follow the instructions below.</p>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"ngsReports"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/smped/ngsReports" class="external-link">ngsReports</a></span><span class="op">)</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p>The paper for the package can be found <a href="https://doi.org/10.1093/bioinformatics/btz937" class="external-link">here</a>. For a detailed usage guide please see <a href="https://bioconductor.org/packages/release/bioc/vignettes/ngsReports/inst/doc/ngsReportsIntroduction.html" class="external-link">here</a>.</p>
</div>
<div class="section level2">
<h2 id="shinyapp">ShinyApp<a class="anchor" aria-label="anchor" href="#shinyapp"></a>
</h2>
<p>A Graphical User Interface (Shiny App) has been developed for interactive inspection of many FastQC reports. The ngsReports shiny app can be installed <a href="https://github.com/UofABioinformaticsHub/shinyNgsReports" class="external-link">here</a>.</p>
</div>
<div class="section level2">
<h2 id="tools-supported-by-category">Tools Supported (By Category)<a class="anchor" aria-label="anchor" href="#tools-supported-by-category"></a>
</h2>
<div class="section level3">
<h3 id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h3>
<ul>
<li>FastQC</li>
<li>fastp</li>
</ul>
</div>
<div class="section level3">
<h3 id="adapter-removal-and-trimming">Adapter removal and trimming<a class="anchor" aria-label="anchor" href="#adapter-removal-and-trimming"></a>
</h3>
<ul>
<li>AdapterRemoval</li>
<li>cutadapt</li>
<li>trimmomatic</li>
<li>fastp</li>
</ul>
</div>
<div class="section level3">
<h3 id="deduplication">Deduplication<a class="anchor" aria-label="anchor" href="#deduplication"></a>
</h3>
<ul>
<li>umi_tools dedup</li>
</ul>
</div>
<div class="section level3">
<h3 id="mapping-and-alignment">Mapping and alignment<a class="anchor" aria-label="anchor" href="#mapping-and-alignment"></a>
</h3>
<ul>
<li>bowtie</li>
<li>bowtie2</li>
<li>hisat2</li>
<li>macs2 callpeak</li>
<li>picard MarkDuplicates</li>
<li>samtools flagstat</li>
<li>STAR</li>
</ul>
</div>
<div class="section level3">
<h3 id="transcriptgene-quantificaiton">Transcript/gene quantificaiton<a class="anchor" aria-label="anchor" href="#transcriptgene-quantificaiton"></a>
</h3>
<ul>
<li>feature counts (Subread)</li>
</ul>
</div>
<div class="section level3">
<h3 id="genome-assembly">Genome assembly<a class="anchor" aria-label="anchor" href="#genome-assembly"></a>
</h3>
<ul>
<li>BUSCO</li>
<li>quast</li>
</ul>
</div>
</div>
</div>
<div class="section level1">
<h1 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h1>
<p>Please cite our <a href="https://doi.org/10.1093/bioinformatics/btz937" class="external-link">paper</a>:</p>
<pre><code>@article{ward2018ngsreports,
    author = {Ward, Christopher M and To, Thu-Hien and Pederson, Stephen M},
    title = "{ngsReports: a Bioconductor package for managing FastQC reports and other NGS related log files}",
    journal = {Bioinformatics},
    year = {2019},
    month = {12},
    doi = {10.1093/bioinformatics/btz937},
    url = {https://doi.org/10.1093/bioinformatics/btz937}
}</code></pre>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/ngsReports" class="external-link">View on Bioconductor</a></li>
<li><a href="https://github.com/smped/ngsReports/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/smped/ngsReports/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing ngsReports</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Stevie Pederson <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-8197-3303" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>
<a href="https://github.com/CMWbio" class="external-link">Christopher Ward</a> <br><small class="roles"> Author </small>  </li>
<li>
<a href="https://github.com/tothuhien" class="external-link">Thu-Hien To</a> <br><small class="roles"> Author </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/smped/ngsReports/actions" class="external-link"><img src="https://github.com/smped/ngsReports/workflows/R-CMD-check-bioc/badge.svg" alt="Build Status"></a></li>
<li><a href="https://shields.io/" class="external-link"><img src="https://img.shields.io/badge/repo%20status-Active-green.svg" alt="Repo Status"></a></li>
<li><a href="https://codecov.io/gh/smped/ngsReports?branch=devel" class="external-link"><img src="https://codecov.io/gh/smped/ngsReports/branch/devel/graph/badge.svg" alt="Codecov test coverage"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Stevie Pederson, <a href="https://github.com/CMWbio" class="external-link">Christopher Ward</a>, <a href="https://github.com/tothuhien" class="external-link">Thu-Hien To</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
