<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ ngsReports</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ngsReports</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.5.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ngsReportsIntroduction.html">An Introduction to ngsReports</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/smped/ngsReports" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="object-classes">Object Classes <a href="#object-classes" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="FastqcData.html">FastqcData()</a></code> </p>
        </td>
        <td><p>The FastqcData Object Class</p></td>
      </tr><tr><td>
          <p><code><a href="FastqcDataList.html">FastqcDataList()</a></code> </p>
        </td>
        <td><p>The FastqcDataList Object Class</p></td>
      </tr><tr><td>
          <p><code><a href="FastpData.html">FastpData()</a></code> </p>
        </td>
        <td><p>The FastpData Object Class</p></td>
      </tr><tr><td>
          <p><code><a href="FastpDataList.html">FastpDataList()</a></code> </p>
        </td>
        <td><p>The FastpDataList Object Class</p></td>
      </tr><tr><td>
          <p><code><a href="PwfCols-class.html">PwfCols-class</a></code> <code><a href="PwfCols-class.html">PwfCols</a></code> </p>
        </td>
        <td><p>The PwfCols class and associated methods</p></td>
      </tr><tr><td>
          <p><code><a href="TheoreticalGC-class.html">TheoreticalGC-class</a></code> </p>
        </td>
        <td><p>The TheoreticalGC Object Class</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="working-with-fastqc-or-fastp-reports">Working With FastQC or Fastp Reports <a href="#working-with-fastqc-or-fastp-reports" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="getModule.html">getModule(<i>&lt;FastqcData&gt;</i>)</a></code> <code><a href="getModule.html">getModule(<i>&lt;FastqcDataList&gt;</i>)</a></code> <code><a href="getModule.html">getModule(<i>&lt;ANY&gt;</i>)</a></code> <code><a href="getModule.html">getModule(<i>&lt;FastpData&gt;</i>)</a></code> <code><a href="getModule.html">getModule(<i>&lt;FastpDataList&gt;</i>)</a></code> </p>
        </td>
        <td><p>Retrieve a given module from a Fastqc* Object</p></td>
      </tr><tr><td>
          <p><code><a href="getSummary.html">getSummary(<i>&lt;.FastqcFile&gt;</i>)</a></code> <code><a href="getSummary.html">getSummary(<i>&lt;ANY&gt;</i>)</a></code> <code><a href="getSummary.html">getSummary(<i>&lt;FastqcData&gt;</i>)</a></code> <code><a href="getSummary.html">getSummary(<i>&lt;FastqcDataList&gt;</i>)</a></code> </p>
        </td>
        <td><p>Get the summary information from Fastqc Files</p></td>
      </tr><tr><td>
          <p><code><a href="plotReadTotals-methods.html">plotReadTotals()</a></code> </p>
        </td>
        <td><p>Draw a barplot of read totals</p></td>
      </tr><tr><td>
          <p><code><a href="readTotals.html">readTotals()</a></code> </p>
        </td>
        <td><p>Get the read totals</p></td>
      </tr><tr><td>
          <p><code><a href="plotSummary-methods.html">plotSummary()</a></code> </p>
        </td>
        <td><p>Plot the PASS/WARN/FAIL information</p></td>
      </tr><tr><td>
          <p><code><a href="plotAdapterContent-methods.html">plotAdapterContent()</a></code> </p>
        </td>
        <td><p>Draw an Adapter Content Plot</p></td>
      </tr><tr><td>
          <p><code><a href="plotBaseQuals-methods.html">plotBaseQuals()</a></code> </p>
        </td>
        <td><p>Plot the Base Qualities for each file</p></td>
      </tr><tr><td>
          <p><code><a href="plotDupLevels-methods.html">plotDupLevels()</a></code> </p>
        </td>
        <td><p>Plot the combined Sequence_Duplication_Levels information</p></td>
      </tr><tr><td>
          <p><code><a href="plotGcContent-methods.html">plotGcContent()</a></code> </p>
        </td>
        <td><p>Plot the Per Sequence GC Content</p></td>
      </tr><tr><td>
          <p><code><a href="plotInsertSize-methods.html">plotInsertSize()</a></code> </p>
        </td>
        <td><p>Plot Insert Size Distributions</p></td>
      </tr><tr><td>
          <p><code><a href="plotKmers-methods.html">plotKmers()</a></code> </p>
        </td>
        <td><p>Plot Overrepresented Kmers</p></td>
      </tr><tr><td>
          <p><code><a href="plotNContent-methods.html">plotNContent()</a></code> </p>
        </td>
        <td><p>Draw an N Content Plot</p></td>
      </tr><tr><td>
          <p><code><a href="plotOverrep-methods.html">plotOverrep()</a></code> </p>
        </td>
        <td><p>Plot a summary of Over-represented Sequences</p></td>
      </tr><tr><td>
          <p><code><a href="plotFastqcPCA-methods.html">plotFastqcPCA()</a></code> </p>
        </td>
        <td><p>Draw a PCA plot for Fast QC modules</p></td>
      </tr><tr><td>
          <p><code><a href="plotSeqContent-methods.html">plotSeqContent()</a></code> </p>
        </td>
        <td><p>Plot the per base content as a heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="plotSeqLengthDistn-methods.html">plotSeqLengthDistn()</a></code> </p>
        </td>
        <td><p>Plot the Sequence Length Distribution</p></td>
      </tr><tr><td>
          <p><code><a href="plotSeqQuals-methods.html">plotSeqQuals()</a></code> </p>
        </td>
        <td><p>Plot the Per Sequence Quality Scores</p></td>
      </tr><tr><td>
          <p><code><a href="writeHtmlReport.html">writeHtmlReport()</a></code> </p>
        </td>
        <td><p>Write an HTML Summary Report</p></td>
      </tr><tr><td>
          <p><code><a href="maxAdapterContent.html">maxAdapterContent()</a></code> </p>
        </td>
        <td><p>Get the maximum Adapter Content</p></td>
      </tr><tr><td>
          <p><code><a href="fqName-methods.html">fqName()</a></code> <code><a href="fqName-methods.html">`fqName&lt;-`()</a></code> </p>
        </td>
        <td><p>Return the Underlying Fastq File Names from Fastqc/Fastp Objects</p></td>
      </tr><tr><td>
          <p><code><a href="fqcVersion.html">fqcVersion(<i>&lt;FastqcData&gt;</i>)</a></code> <code><a href="fqcVersion.html">fqcVersion(<i>&lt;FastqcDataList&gt;</i>)</a></code> <code><a href="fqcVersion.html">fqcVersion(<i>&lt;ANY&gt;</i>)</a></code> </p>
        </td>
        <td><p>Get the FASTQC version</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="working-with-other-ngs-tools">Working With Other NGS Tools <a href="#working-with-other-ngs-tools" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="importNgsLogs.html">importNgsLogs()</a></code> </p>
        </td>
        <td><p>Import Various NGS-related log files</p></td>
      </tr><tr><td>
          <p><code><a href="importSJ.html">importSJ()</a></code> </p>
        </td>
        <td><p>Import STAR Splice Junctions</p></td>
      </tr><tr><td>
          <p><code><a href="plotAlignmentSummary.html">plotAlignmentSummary()</a></code> </p>
        </td>
        <td><p>Plot a summary of alignments</p></td>
      </tr><tr><td>
          <p><code><a href="plotAssemblyStats.html">plotAssemblyStats()</a></code> </p>
        </td>
        <td><p>Plot a summary of assembly logs</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="utility-functions">Utility Functions <a href="#utility-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="estGcDistn.html">estGcDistn()</a></code> </p>
        </td>
        <td><p>Estimate a GC Content Distribution From Sequences</p></td>
      </tr><tr><td>
          <p><code><a href="gcAvail.html">gcAvail()</a></code> </p>
        </td>
        <td><p>List Genomes or Transcriptomes with Theoretical GC Content</p></td>
      </tr><tr><td>
          <p><code><a href="getGC.html">getGC()</a></code> </p>
        </td>
        <td><p>Get Theoretical GC content</p></td>
      </tr><tr><td>
          <p><code><a href="getColours-methods.html">getColours(<i>&lt;PwfCols&gt;</i>)</a></code> <code><a href="getColours-methods.html">setColours(<i>&lt;PwfCols&gt;</i>)</a></code> <code><a href="getColours-methods.html">setAlpha(<i>&lt;PwfCols&gt;</i>)</a></code> </p>
        </td>
        <td><p>Work with objects of class PwfCols</p></td>
      </tr><tr><td>
          <p><code><a href="overRep2Fasta-methods.html">overRep2Fasta()</a></code> </p>
        </td>
        <td><p>Write fasta of Over-Represented sequences.</p></td>
      </tr><tr><td>
          <p><code><a href="isCompressed.html">isCompressed()</a></code> </p>
        </td>
        <td><p>Check to see if a file is compressed</p></td>
      </tr><tr><td>
          <p><code><a href="path.html">path(<i>&lt;.FastqcFile&gt;</i>)</a></code> <code><a href="path.html">path(<i>&lt;FastqcData&gt;</i>)</a></code> <code><a href="path.html">path(<i>&lt;FastqcDataList&gt;</i>)</a></code> <code><a href="path.html">path(<i>&lt;.FastpFile&gt;</i>)</a></code> <code><a href="path.html">path(<i>&lt;FastpData&gt;</i>)</a></code> <code><a href="path.html">path(<i>&lt;FastpDataList&gt;</i>)</a></code> </p>
        </td>
        <td><p>Return the File Paths from an object</p></td>
      </tr><tr><td>
          <p><code><a href="mData.html">mData()</a></code> </p>
        </td>
        <td><p>Extract Metadata for TheoreticalGC objects</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="data-objects">Data Objects <a href="#data-objects" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="gcTheoretical.html">gcTheoretical</a></code> </p>
        </td>
        <td><p>Theoretical GC content</p></td>
      </tr><tr><td>
          <p><code><a href="pwf.html">pwf</a></code> </p>
        </td>
        <td><p>Colours for PASS/WARN/FAIL</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="all-methods">All Methods <a href="#all-methods" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="estGcDistn.html">estGcDistn()</a></code> </p>
        </td>
        <td><p>Estimate a GC Content Distribution From Sequences</p></td>
      </tr><tr><td>
          <p><code><a href="extract-methods.html">`[`(<i>&lt;FastqcDataList&gt;</i>,<i>&lt;numeric&gt;</i>,<i>&lt;missing&gt;</i>)</a></code> <code><a href="extract-methods.html">`[`(<i>&lt;FastqcDataList&gt;</i>,<i>&lt;character&gt;</i>,<i>&lt;missing&gt;</i>)</a></code> <code><a href="extract-methods.html">`[`(<i>&lt;FastqcDataList&gt;</i>,<i>&lt;logical&gt;</i>,<i>&lt;missing&gt;</i>)</a></code> <code><a href="extract-methods.html">`[`(<i>&lt;FastqcDataList&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;missing&gt;</i>)</a></code> <code><a href="extract-methods.html">`[`(<i>&lt;FastpDataList&gt;</i>,<i>&lt;numeric&gt;</i>,<i>&lt;missing&gt;</i>)</a></code> <code><a href="extract-methods.html">`[`(<i>&lt;FastpDataList&gt;</i>,<i>&lt;character&gt;</i>,<i>&lt;missing&gt;</i>)</a></code> <code><a href="extract-methods.html">`[`(<i>&lt;FastpDataList&gt;</i>,<i>&lt;logical&gt;</i>,<i>&lt;missing&gt;</i>)</a></code> <code><a href="extract-methods.html">`[`(<i>&lt;FastpDataList&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;missing&gt;</i>)</a></code> </p>
        </td>
        <td><p>Extract Elements</p></td>
      </tr><tr><td>
          <p><code><a href="fqName-methods.html">fqName()</a></code> <code><a href="fqName-methods.html">`fqName&lt;-`()</a></code> </p>
        </td>
        <td><p>Return the Underlying Fastq File Names from Fastqc/Fastp Objects</p></td>
      </tr><tr><td>
          <p><code><a href="fqcVersion.html">fqcVersion(<i>&lt;FastqcData&gt;</i>)</a></code> <code><a href="fqcVersion.html">fqcVersion(<i>&lt;FastqcDataList&gt;</i>)</a></code> <code><a href="fqcVersion.html">fqcVersion(<i>&lt;ANY&gt;</i>)</a></code> </p>
        </td>
        <td><p>Get the FASTQC version</p></td>
      </tr><tr><td>
          <p><code><a href="gcAvail.html">gcAvail()</a></code> </p>
        </td>
        <td><p>List Genomes or Transcriptomes with Theoretical GC Content</p></td>
      </tr><tr><td>
          <p><code><a href="getColours-methods.html">getColours(<i>&lt;PwfCols&gt;</i>)</a></code> <code><a href="getColours-methods.html">setColours(<i>&lt;PwfCols&gt;</i>)</a></code> <code><a href="getColours-methods.html">setAlpha(<i>&lt;PwfCols&gt;</i>)</a></code> </p>
        </td>
        <td><p>Work with objects of class PwfCols</p></td>
      </tr><tr><td>
          <p><code><a href="getGC.html">getGC()</a></code> </p>
        </td>
        <td><p>Get Theoretical GC content</p></td>
      </tr><tr><td>
          <p><code><a href="getModule.html">getModule(<i>&lt;FastqcData&gt;</i>)</a></code> <code><a href="getModule.html">getModule(<i>&lt;FastqcDataList&gt;</i>)</a></code> <code><a href="getModule.html">getModule(<i>&lt;ANY&gt;</i>)</a></code> <code><a href="getModule.html">getModule(<i>&lt;FastpData&gt;</i>)</a></code> <code><a href="getModule.html">getModule(<i>&lt;FastpDataList&gt;</i>)</a></code> </p>
        </td>
        <td><p>Retrieve a given module from a Fastqc* Object</p></td>
      </tr><tr><td>
          <p><code><a href="getSummary.html">getSummary(<i>&lt;.FastqcFile&gt;</i>)</a></code> <code><a href="getSummary.html">getSummary(<i>&lt;ANY&gt;</i>)</a></code> <code><a href="getSummary.html">getSummary(<i>&lt;FastqcData&gt;</i>)</a></code> <code><a href="getSummary.html">getSummary(<i>&lt;FastqcDataList&gt;</i>)</a></code> </p>
        </td>
        <td><p>Get the summary information from Fastqc Files</p></td>
      </tr><tr><td>
          <p><code><a href="mData.html">mData()</a></code> </p>
        </td>
        <td><p>Extract Metadata for TheoreticalGC objects</p></td>
      </tr><tr><td>
          <p><code><a href="overRep2Fasta-methods.html">overRep2Fasta()</a></code> </p>
        </td>
        <td><p>Write fasta of Over-Represented sequences.</p></td>
      </tr><tr><td>
          <p><code><a href="path.html">path(<i>&lt;.FastqcFile&gt;</i>)</a></code> <code><a href="path.html">path(<i>&lt;FastqcData&gt;</i>)</a></code> <code><a href="path.html">path(<i>&lt;FastqcDataList&gt;</i>)</a></code> <code><a href="path.html">path(<i>&lt;.FastpFile&gt;</i>)</a></code> <code><a href="path.html">path(<i>&lt;FastpData&gt;</i>)</a></code> <code><a href="path.html">path(<i>&lt;FastpDataList&gt;</i>)</a></code> </p>
        </td>
        <td><p>Return the File Paths from an object</p></td>
      </tr><tr><td>
          <p><code><a href="plotAdapterContent-methods.html">plotAdapterContent()</a></code> </p>
        </td>
        <td><p>Draw an Adapter Content Plot</p></td>
      </tr><tr><td>
          <p><code><a href="plotBaseQuals-methods.html">plotBaseQuals()</a></code> </p>
        </td>
        <td><p>Plot the Base Qualities for each file</p></td>
      </tr><tr><td>
          <p><code><a href="plotDupLevels-methods.html">plotDupLevels()</a></code> </p>
        </td>
        <td><p>Plot the combined Sequence_Duplication_Levels information</p></td>
      </tr><tr><td>
          <p><code><a href="plotFastqcPCA-methods.html">plotFastqcPCA()</a></code> </p>
        </td>
        <td><p>Draw a PCA plot for Fast QC modules</p></td>
      </tr><tr><td>
          <p><code><a href="plotGcContent-methods.html">plotGcContent()</a></code> </p>
        </td>
        <td><p>Plot the Per Sequence GC Content</p></td>
      </tr><tr><td>
          <p><code><a href="plotInsertSize-methods.html">plotInsertSize()</a></code> </p>
        </td>
        <td><p>Plot Insert Size Distributions</p></td>
      </tr><tr><td>
          <p><code><a href="plotKmers-methods.html">plotKmers()</a></code> </p>
        </td>
        <td><p>Plot Overrepresented Kmers</p></td>
      </tr><tr><td>
          <p><code><a href="plotNContent-methods.html">plotNContent()</a></code> </p>
        </td>
        <td><p>Draw an N Content Plot</p></td>
      </tr><tr><td>
          <p><code><a href="plotOverrep-methods.html">plotOverrep()</a></code> </p>
        </td>
        <td><p>Plot a summary of Over-represented Sequences</p></td>
      </tr><tr><td>
          <p><code><a href="plotReadTotals-methods.html">plotReadTotals()</a></code> </p>
        </td>
        <td><p>Draw a barplot of read totals</p></td>
      </tr><tr><td>
          <p><code><a href="plotSeqContent-methods.html">plotSeqContent()</a></code> </p>
        </td>
        <td><p>Plot the per base content as a heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="plotSeqLengthDistn-methods.html">plotSeqLengthDistn()</a></code> </p>
        </td>
        <td><p>Plot the Sequence Length Distribution</p></td>
      </tr><tr><td>
          <p><code><a href="plotSeqQuals-methods.html">plotSeqQuals()</a></code> </p>
        </td>
        <td><p>Plot the Per Sequence Quality Scores</p></td>
      </tr><tr><td>
          <p><code><a href="plotSummary-methods.html">plotSummary()</a></code> </p>
        </td>
        <td><p>Plot the PASS/WARN/FAIL information</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Stevie Pederson, <a href="https://github.com/CMWbio" class="external-link">Christopher Ward</a>, <a href="https://github.com/tothuhien" class="external-link">Thu-Hien To</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

