<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Plot the Per Sequence GC Content — plotGcContent • ngsReports</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Plot the Per Sequence GC Content — plotGcContent"><meta property="og:description" content="Plot the Per Sequence GC Content for a set of FASTQC files"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ngsReports</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.3.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ngsReportsIntroduction.html">An Introduction to ngsReports</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/smped/ngsReports" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot the Per Sequence GC Content</h1>
    <small class="dont-index">Source: <a href="https://github.com/smped/ngsReports/blob/HEAD/R/plotGcContent.R" class="external-link"><code>R/plotGcContent.R</code></a></small>
    <div class="hidden name"><code>plotGcContent-methods.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Plot the Per Sequence GC Content for a set of FASTQC files</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotGcContent</span><span class="op">(</span><span class="va">x</span>, usePlotly <span class="op">=</span> <span class="cn">FALSE</span>, <span class="va">labels</span>, pattern <span class="op">=</span> <span class="st">".(fast|fq|bam).*"</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for ANY</span></span>
<span><span class="fu">plotGcContent</span><span class="op">(</span><span class="va">x</span>, usePlotly <span class="op">=</span> <span class="cn">FALSE</span>, <span class="va">labels</span>, pattern <span class="op">=</span> <span class="st">".(fast|fq|bam).*"</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for FastqcData</span></span>
<span><span class="fu">plotGcContent</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  usePlotly <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">labels</span>,</span>
<span>  pattern <span class="op">=</span> <span class="st">".(fast|fq|bam).*"</span>,</span>
<span>  theoreticalGC <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  gcType <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Genome"</span>, <span class="st">"Transcriptome"</span><span class="op">)</span>,</span>
<span>  species <span class="op">=</span> <span class="st">"Hsapiens"</span>,</span>
<span>  <span class="va">GCobject</span>,</span>
<span>  plotlyLegend <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">Fastafile</span>,</span>
<span>  n <span class="op">=</span> <span class="fl">1e+06</span>,</span>
<span>  counts <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  scaleColour <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  lineCols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"red3"</span>, <span class="st">"black"</span><span class="op">)</span>,</span>
<span>  linetype <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  linewidth <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for FastqcDataList</span></span>
<span><span class="fu">plotGcContent</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  usePlotly <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">labels</span>,</span>
<span>  pattern <span class="op">=</span> <span class="st">".(fast|fq|bam).*"</span>,</span>
<span>  theoreticalGC <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  gcType <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Genome"</span>, <span class="st">"Transcriptome"</span><span class="op">)</span>,</span>
<span>  species <span class="op">=</span> <span class="st">"Hsapiens"</span>,</span>
<span>  <span class="va">GCobject</span>,</span>
<span>  <span class="va">Fastafile</span>,</span>
<span>  n <span class="op">=</span> <span class="fl">1e+06</span>,</span>
<span>  plotType <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"heatmap"</span>, <span class="st">"line"</span>, <span class="st">"cdf"</span><span class="op">)</span>,</span>
<span>  cluster <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  dendrogram <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  heat_w <span class="op">=</span> <span class="fl">8</span>,</span>
<span>  <span class="va">pwfCols</span>,</span>
<span>  showPwf <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  scaleFill <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  scaleColour <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plotlyLegend <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  lineCols <span class="op">=</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fl">12</span>, <span class="st">"Paired"</span><span class="op">)</span>,</span>
<span>  linetype <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  linewidth <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for FastpData</span></span>
<span><span class="fu">plotGcContent</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  usePlotly <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">labels</span>,</span>
<span>  pattern <span class="op">=</span> <span class="st">".(fast|fq|bam).*"</span>,</span>
<span>  theoreticalGC <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  gcType <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Genome"</span>, <span class="st">"Transcriptome"</span><span class="op">)</span>,</span>
<span>  species <span class="op">=</span> <span class="st">"Hsapiens"</span>,</span>
<span>  <span class="va">GCobject</span>,</span>
<span>  <span class="va">Fastafile</span>,</span>
<span>  n <span class="op">=</span> <span class="fl">1e+06</span>,</span>
<span>  plotType <span class="op">=</span> <span class="st">"bar"</span>,</span>
<span>  scaleFill <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plotlyLegend <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  plotTheme <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for FastpDataList</span></span>
<span><span class="fu">plotGcContent</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  usePlotly <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">labels</span>,</span>
<span>  pattern <span class="op">=</span> <span class="st">".(fast|fq|bam).*"</span>,</span>
<span>  theoreticalGC <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  gcType <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Genome"</span>, <span class="st">"Transcriptome"</span><span class="op">)</span>,</span>
<span>  species <span class="op">=</span> <span class="st">"Hsapiens"</span>,</span>
<span>  <span class="va">GCobject</span>,</span>
<span>  <span class="va">Fastafile</span>,</span>
<span>  n <span class="op">=</span> <span class="fl">1e+06</span>,</span>
<span>  plotType <span class="op">=</span> <span class="st">"bar"</span>,</span>
<span>  scaleFill <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plotTheme <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  plotlyLegend <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>x</dt>
<dd><p>Can be a <code>FastqcData</code>, <code>FastqcDataList</code> or character
vector of file paths</p></dd>


<dt>usePlotly</dt>
<dd><p><code>logical</code> Default <code>FALSE</code> will render using
ggplot. If <code>TRUE</code> plot will be rendered with plotly</p></dd>


<dt>labels</dt>
<dd><p>An optional named vector of labels for the file names.</p></dd>


<dt>pattern</dt>
<dd><p>Pattern to remove from the end of filenames</p></dd>


<dt>...</dt>
<dd><p>Used to pass various potting parameters to themes and geoms.</p></dd>


<dt>theoreticalGC</dt>
<dd><p><code>logical</code> default is <code>FALSE</code> to give the true
GC content, set to <code>TRUE</code> to normalize values of GC_Content by the
theoretical values using <code><a href="gcTheoretical.html">gcTheoretical()</a></code>. <code>species</code> must be
specified. For Fastqc* objects, the entire distributions will be used,
wheras for the Fastp* objects, only the expected mean value is shown as a
horizontal line</p></dd>


<dt>gcType</dt>
<dd><p><code>character</code> Select type of data to normalize GC
content against. Accepts either "Genome" (default) or "Transcriptome".</p></dd>


<dt>species</dt>
<dd><p><code>character</code> if <code>gcTheory</code> is <code>TRUE</code> it must be
accompanied by a species. Species currently supported can be obtained using
<code>mData(gcTheoretical)</code></p></dd>


<dt>GCobject</dt>
<dd><p>an object of class GCTheoretical.
Defaults to the gcTheoretical object supplied with the package</p></dd>


<dt>plotlyLegend</dt>
<dd><p>logical(1) Show legend on interactive line plots</p></dd>


<dt>Fastafile</dt>
<dd><p>a fasta file contains DNA sequences to generate theoretical
GC content</p></dd>


<dt>n</dt>
<dd><p>number of simulated reads to generate theoretical GC content from
<code>Fastafile</code></p></dd>


<dt>counts</dt>
<dd><p><code>logical</code>. Plot the counts from each file if
<code>counts = TRUE</code>, otherwise frequencies will be plotted.
Ignored if calling the function on a FastqcDataList.</p></dd>


<dt>scaleColour</dt>
<dd><p>ggplot2 scale for line colours</p></dd>


<dt>lineCols, linetype, linewidth</dt>
<dd><p>Line colour type and width for observed
and theoretical GC lines</p></dd>


<dt>plotType</dt>
<dd><p>Takes values "line", "heatmap" or "cdf"</p></dd>


<dt>cluster</dt>
<dd><p><code>logical</code> default <code>FALSE</code>. If set to <code>TRUE</code>,
fastqc data will be clustered using hierarchical clustering</p></dd>


<dt>dendrogram</dt>
<dd><p><code>logical</code> redundant if <code>cluster</code> is <code>FALSE</code>
if both <code>cluster</code> and <code>dendrogram</code> are specified as <code>TRUE</code>
then the dendrogram  will be displayed.</p></dd>


<dt>heat_w</dt>
<dd><p>Relative width of any heatmap plot components</p></dd>


<dt>pwfCols</dt>
<dd><p>Object of class <code><a href="PwfCols-class.html">PwfCols()</a></code> to give colours for
pass, warning, and fail values in plot</p></dd>


<dt>showPwf</dt>
<dd><p>logical(1) Show Pwf Status on the plot</p></dd>


<dt>scaleFill</dt>
<dd><p>ggplot2 scale for filling heatmap cells or bars</p></dd>


<dt>plotTheme</dt>
<dd><p><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a> object</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A ggplot2 or plotly object</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Makes plots for GC_Content.
When applied to a single FastqcData object a simple line plot will be drawn,
with Theoretical GC content overlaid if desired.</p>
<p>When applied to multiple FastQC reports, the density at each GC content bin
can be shown as a heatmap by setting <code>theoreticalGC = FALSE</code>. By
default the difference in observed and expected theoretical GC is shown.
Species and genome/transcriptome should also be set if utilising the
theoretical GC content.</p>
<p>As an alternative to a heatmap, a series of overlaid distributions can be
shown by setting <code>plotType = "line"</code>.</p>
<p>Can produce a static ggplot2 object or an interactive plotly object.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Get the files included with the package</span></span></span>
<span class="r-in"><span><span class="va">packageDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"ngsReports"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">packageDir</span>, pattern <span class="op">=</span> <span class="st">"fastqc.zip"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Load the FASTQC data as a FastqcDataList object</span></span></span>
<span class="r-in"><span><span class="va">fdl</span> <span class="op">&lt;-</span> <span class="fu"><a href="FastqcDataList.html">FastqcDataList</a></span><span class="op">(</span><span class="va">fl</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># The default plot for a FastqcDataList</span></span></span>
<span class="r-in"><span><span class="fu">plotGcContent</span><span class="op">(</span><span class="va">fdl</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plotGcContent-methods-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Plot a single FastqcData object</span></span></span>
<span class="r-in"><span><span class="fu">plotGcContent</span><span class="op">(</span><span class="va">fdl</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plotGcContent-methods-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Stevie Pederson, <a href="https://github.com/CMWbio" class="external-link">Christopher Ward</a>, <a href="https://github.com/tothuhien" class="external-link">Thu-Hien To</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

