<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>An Introduction To ngsReports • ngsReports</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="An Introduction To ngsReports">
<meta property="og:description" content="ngsReports">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ngsReports</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.5.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ngsReportsIntroduction.html">An Introduction to ngsReports</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/smped/ngsReports" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>An Introduction To ngsReports</h1>
                        <h4 data-toc-skip class="author">Christopher
Ward</h4>
            <address class="author_afil">
      School of Biological Sciences, University of
Adelaide<br><h4 data-toc-skip class="author">Hien
To</h4>
            <address class="author_afil">
      Bioinformatics Hub, University of
Adelaide<br><h4 data-toc-skip class="author">Stevie
Pederson</h4>
            <address class="author_afil">
      Black Ochre Data Laboratories, Telethon Kids Institute, Adelaide,
AustraliaDame Roma Mitchell Cancer Researc Laboratories, University of
AdelaideJohn Curtin School of Medical Research, Australian National
University<br><a class="author_email" href="mailto:#"></a><a href="mailto:stephen.pederson.au@gmail.com" class="email">stephen.pederson.au@gmail.com</a>
      </address>
                  
      
      <small class="dont-index">Source: <a href="https://github.com/smped/ngsReports/blob/HEAD/vignettes/ngsReportsIntroduction.Rmd" class="external-link"><code>vignettes/ngsReportsIntroduction.Rmd</code></a></small>
      <div class="hidden name"><code>ngsReportsIntroduction.Rmd</code></div>

    </address>
</address>
</div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The package <code>ngsReports</code> is designed to bolt into data
processing pipelines and produce combined plots for multiple FastQC
reports generated across an entire set of libraries or samples. The
primary functionality of the package is parsing FastQC reports, with
import methods also implemented for log files produced by tools as as
<code>STAR</code>, <code>hisat2</code> and others. In addition to
parsing files, default plotting methods are implemented. Plots applied
to a single file will replicate the default plots from
<code>FastQC</code><a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>, whilst methods applied to multiple FastQC
reports summarise these and produce a series of custom plots.</p>
<p>Plots are produced as standard <em><a href="https://CRAN.R-project.org/package=ggplot2" class="external-link">ggplot2</a></em>
objects, with an interactive option available using <em><a href="https://CRAN.R-project.org/package=plotly" class="external-link">plotly</a></em>. As
well as custom summary plots, tables of read counts and the like can
also be easily generated.</p>
</div>
<div class="section level2">
<h2 id="basic-usage">Basic Usage<a class="anchor" aria-label="anchor" href="#basic-usage"></a>
</h2>
<div class="section level3">
<h3 id="using-the-shiny-app">Using the Shiny App<a class="anchor" aria-label="anchor" href="#using-the-shiny-app"></a>
</h3>
<p>In addition to the usage demonstrated below, a <code>shiny</code> app
has been developed for interactive viewing of FastQC reports. This can
be installed using:</p>
<pre><code><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"UofABioinformaticsHub/shinyNgsreports"</span><span class="op">)</span></span></code></pre>
<p>A vignette for this app will be installed with the
<code>shinyNgsreports</code> package.</p>
</div>
<div class="section level3">
<h3 id="the-default-report">The default report<a class="anchor" aria-label="anchor" href="#the-default-report"></a>
</h3>
<p>In it’s simplest form, a default summary report can be generated
simply by specifying a directory containing the output from FastQC and
calling the function <code><a href="../reference/writeHtmlReport.html">writeHtmlReport()</a></code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/smped/ngsReports" class="external-link">ngsReports</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="va">fileDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"path"</span>, <span class="st">"to"</span>, <span class="st">"your"</span>, <span class="st">"FastQC"</span>, <span class="st">"Reports"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/writeHtmlReport.html">writeHtmlReport</a></span><span class="op">(</span><span class="va">fileDir</span><span class="op">)</span></span></code></pre>
<p>This function will transfer the default template to the provided
directory and produce a single <code>.html</code> file containing
interactive summary plots of any FastQC output found in the directory.
FastQC output can be <code>*fastqc.zip</code> files or the same files
extracted as individual directories.</p>
<p>The default template is provided as
<code>ngsReports_Fastqc.Rmd</code> in the package directory . This
template can be easily modified and supplied as an alternate template to
the above function using your modified file as the template RMarkdown
file.</p>
<pre><code><span><span class="va">altTemplate</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"path"</span>, <span class="st">"to"</span>, <span class="st">"your"</span>, <span class="st">"new"</span>, <span class="st">"template.Rmd"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/writeHtmlReport.html">writeHtmlReport</a></span><span class="op">(</span><span class="va">fileDir</span>, template <span class="op">=</span> <span class="va">altTemplate</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="advanced-usage">Advanced Usage<a class="anchor" aria-label="anchor" href="#advanced-usage"></a>
</h2>
<div class="section level3">
<h3 id="classes-defined-in-the-package">Classes Defined in the Package<a class="anchor" aria-label="anchor" href="#classes-defined-in-the-package"></a>
</h3>
<p>The package <code>ngsReports</code> introduces two main
<code>S4</code> classes:</p>
<ul>
<li>
<code>FastqcData</code> &amp; <code>FastqcDataList</code>
</li>
</ul>
<p><code>FastqcData</code> objects hold the <em>parsed data from a
single report</em> as generated by the stand-alone tool
<code>FastQC</code>. These are then extended into lists for <em>more
than one file</em> as a <code>FastqcDataList</code>. For most users, the
primary class of interest will be the <code>FastqcDataList</code>.</p>
</div>
<div class="section level3">
<h3 id="loading-fastqc-data-into-r">Loading FastQC Data Into <code>R</code><a class="anchor" aria-label="anchor" href="#loading-fastqc-data-into-r"></a>
</h3>
<p>To load a set of <code>FastQC</code> reports into <code>R</code> as a
<code>FastqcDataList</code>, specify the vector of file paths, then call
the function <code><a href="../reference/FastqcDataList.html">FastqcDataList()</a></code>. In the rare case you’d like
an individual file, this can be performed by calling
<code><a href="../reference/FastqcData.html">FastqcData()</a></code> on an individual file, or subsetting the
output from <code><a href="../reference/FastqcDataList.html">FastqcDataList()</a></code> using the <code>[[]]</code>
operator as with any list object.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fileDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"ngsReports"</span><span class="op">)</span></span>
<span><span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">fileDir</span>, pattern <span class="op">=</span> <span class="st">"fastqc.zip$"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">fdl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FastqcDataList.html">FastqcDataList</a></span><span class="op">(</span><span class="va">files</span><span class="op">)</span></span></code></pre></div>
<p>From here, all FastQC modules can be obtained as a
<code>tibble</code> (i.e. <code>data.frame</code>) using the function
<code><a href="../reference/getModule.html">getModule()</a></code> and choosing one of the following modules:</p>
<ul>
<li>
<code>Summary</code> (The PASS/WARN/FAIL status for each
module)</li>
<li><code>Basic_Statistics</code></li>
<li><code>Per_base_sequence_quality</code></li>
<li><code>Per_sequence_quality_scores</code></li>
<li><code>Per_base_sequence_content</code></li>
<li><code>Per_sequence_GC_content</code></li>
<li><code>Per_base_N_content</code></li>
<li><code>Sequence_Length_Distribution</code></li>
<li><code>Sequence_Duplication_Levels</code></li>
<li><code>Overrepresented_sequences</code></li>
<li><code>Adapter_Content</code></li>
<li><code>Kmer_Content</code></li>
<li><code>Per_tile_sequence_quality</code></li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getModule.html">getModule</a></span><span class="op">(</span><span class="va">fdl</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="st">"Summary"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 12 × 3</span></span></span>
<span><span class="co">##    Filename      Status Category                    </span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                       </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> ATTG_R1.fastq PASS   Basic Statistics            </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> ATTG_R1.fastq FAIL   Per base sequence quality   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> ATTG_R1.fastq WARN   Per tile sequence quality   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> ATTG_R1.fastq PASS   Per sequence quality scores </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> ATTG_R1.fastq FAIL   Per base sequence content   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> ATTG_R1.fastq FAIL   Per sequence GC content     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> ATTG_R1.fastq PASS   Per base N content          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> ATTG_R1.fastq PASS   Sequence Length Distribution</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> ATTG_R1.fastq FAIL   Sequence Duplication Levels </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> ATTG_R1.fastq FAIL   Overrepresented sequences   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">11</span> ATTG_R1.fastq FAIL   Adapter Content             </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">12</span> ATTG_R1.fastq FAIL   Kmer Content</span></span></code></pre>
<p>Capitalisation and spelling of these module names follows the default
patterns from FastQC reports with spaces replaced by underscores. One
additional module is available and taken directly from the text within
the supplied reports</p>
<ul>
<li><code>Total_Duplicated_Percentage</code></li>
</ul>
<p>In addition, the read totals for each file in the library can be
obtained using <code><a href="../reference/readTotals.html">readTotals()</a></code>, which can be easily used to
make a table of read totals. This essentially just returns the first two
columns from <code>getModule(x, "Basic_Statistics")</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reads</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readTotals.html">readTotals</a></span><span class="op">(</span><span class="va">fdl</span><span class="op">)</span></span></code></pre></div>
<p>The packages <code>dplyr</code> and <code>pander</code> can also be
extremely useful for manipulating and displaying imported data. To show
only the R1 read totals, you could do the following</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rapporter.github.io/pander/" class="external-link">pander</a></span><span class="op">)</span></span>
<span><span class="va">reads</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"R1"</span>, <span class="va">Filename</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/pander/man/pander.html" class="external-link">pander</a></span><span class="op">(</span></span>
<span>        big.mark <span class="op">=</span> <span class="st">","</span>,</span>
<span>        caption <span class="op">=</span> <span class="st">"Read totals from R1 libraries"</span>, </span>
<span>        justify <span class="op">=</span> <span class="st">"lr"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<table style="width:47%;" class="table">
<caption>Read totals from R1 libraries</caption>
<colgroup>
<col width="22%">
<col width="25%">
</colgroup>
<thead><tr class="header">
<th align="left">Filename</th>
<th align="right">Total_Sequences</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ATTG_R1.fastq</td>
<td align="right">24,978</td>
</tr>
<tr class="even">
<td align="left">CCGC_R1.fastq</td>
<td align="right">22,269</td>
</tr>
<tr class="odd">
<td align="left">GACC_R1.fastq</td>
<td align="right">10,287</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="generating-plots-for-one-or-more-fastqc-files">Generating Plots For One or More Fastqc Files<a class="anchor" aria-label="anchor" href="#generating-plots-for-one-or-more-fastqc-files"></a>
</h3>
<p>Plots created from a single <code>FastqcData</code> object will
resemble those generated by the <code>FastQC</code> tool, whilst those
created from a <code>FastqcDataList</code> will be combined summaries
across a library of files. In addition, all plots are able to be
generated as interactive plots using the argument
<code>usePlotly = TRUE</code>.</p>
<p>All FastQC modules have been enabled for plotting using default
<code>S4</code> dispatch, with the exception of
<code>Per_tile_sequence_quality</code>.</p>
<div class="section level4">
<h4 id="inspecting-the-passwarnfail-status-of-each-module">Inspecting the PASS/WARN/FAIL Status of each module<a class="anchor" aria-label="anchor" href="#inspecting-the-passwarnfail-status-of-each-module"></a>
</h4>
<p>The simplest of the plots is to summarise the
<code>PASS/WARN/FAIL</code> flags as produced by <code>FastQC</code> for
each module. This plot can be simply generated using
<code><a href="../reference/plotSummary-methods.html">plotSummary()</a></code></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSummary-methods.html">plotSummary</a></span><span class="op">(</span><span class="va">fdl</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="ngsReportsIntroduction_files/figure-html/plotSummary-1.png" alt="Default summary of FastQC flags." width="700"><p class="caption">
Default summary of FastQC flags.
</p>
</div>
</div>
<div class="section level4">
<h4 id="visualising-read-totals">Visualising Read Totals<a class="anchor" aria-label="anchor" href="#visualising-read-totals"></a>
</h4>
<p>The next most informative plot may be to summarise the total numbers
of reads in each associated Fastq file. By default, the number of
duplicated sequences from the <code>Total_Duplicated_Percentage</code>
module are shown, but this can be disabled by setting
<code>duplicated = FALSE</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotReadTotals-methods.html">plotReadTotals</a></span><span class="op">(</span><span class="va">fdl</span><span class="op">)</span></span></code></pre></div>
<p><img src="ngsReportsIntroduction_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>As these are <code>ggplot2</code> objects, the output can be modified
easily using conventional <code>ggplot2</code> syntax. Here we’ll move
the legend to the top right as an example.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotReadTotals-methods.html">plotReadTotals</a></span><span class="op">(</span><span class="va">fdl</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>        legend.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>, </span>
<span>        legend.justification <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>        legend.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_rect</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p><img src="ngsReportsIntroduction_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="per-base-sequence-qualities">Per Base Sequence Qualities<a class="anchor" aria-label="anchor" href="#per-base-sequence-qualities"></a>
</h4>
<p>Turning to the <code>Per base sequence quality</code> scores is the
next most common step for most researchers, and these can be obtained
for an individual file by selecting this as an element
(i.e. <code>FastqcData</code> object ) of the main
<code>FastqcDataList</code> object. This plot replicates the default
plots from a FastQC report.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotBaseQuals-methods.html">plotBaseQuals</a></span><span class="op">(</span><span class="va">fdl</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="ngsReportsIntroduction_files/figure-html/unnamed-chunk-9-1.png" alt="Example showing the Per_base_sequence_quality plot for a single FastqcData object." width="700"><p class="caption">
Example showing the Per_base_sequence_quality plot for a single
FastqcData object.
</p>
</div>
<p>When working with multiple FastQC reports, these are summarised as a
heatmap using the mean quality score at each position.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotBaseQuals-methods.html">plotBaseQuals</a></span><span class="op">(</span><span class="va">fdl</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="ngsReportsIntroduction_files/figure-html/unnamed-chunk-10-1.png" alt="Example showing the Mean Per Base Squence Qualities for a set of FastQC reports." width="700"><p class="caption">
Example showing the Mean Per Base Squence Qualities for a set of FastQC
reports.
</p>
</div>
<p>Boxplots of any combinations can also be drawn from a
<code>FastqcDataList</code> by setting the argument
<code>plotType = "boxplot"</code>. However, this may be not suitable for
datasets with a large number of libraries.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotBaseQuals-methods.html">plotBaseQuals</a></span><span class="op">(</span><span class="va">fdl</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>, plotType <span class="op">=</span> <span class="st">"boxplot"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ngsReportsIntroduction_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="mean-sequence-quality-per-read">Mean Sequence Quality Per Read<a class="anchor" aria-label="anchor" href="#mean-sequence-quality-per-read"></a>
</h4>
<p>Similarly, the Mean Sequence Quality Per Read plot can be generated
to replicate plots from a FastQC report by selecting the individual file
from the <code>FastqcDataList</code> object.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSeqQuals-methods.html">plotSeqQuals</a></span><span class="op">(</span><span class="va">fdl</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="ngsReportsIntroduction_files/figure-html/unnamed-chunk-12-1.png" alt="Example plot showing Per_sequence_quality_scores for an individual file." width="700"><p class="caption">
Example plot showing Per_sequence_quality_scores for an individual file.
</p>
</div>
<p>A heatmap of mean sequence qualities can be generated when inspecting
multiple reports.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSeqQuals-methods.html">plotSeqQuals</a></span><span class="op">(</span><span class="va">fdl</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="ngsReportsIntroduction_files/figure-html/unnamed-chunk-13-1.png" alt="Example heatmaps showing Per_sequence_quality_scores for a set of files." width="700"><p class="caption">
Example heatmaps showing Per_sequence_quality_scores for a set of files.
</p>
</div>
<p>An alternative view may be to plot these as overlaid lines, which can
be simply done by setting <code>plotType = "line"</code>. Again,
discretion should be shown when choosing this option for many
samples.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">r2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"R2"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">fdl</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotSeqQuals-methods.html">plotSeqQuals</a></span><span class="op">(</span><span class="va">fdl</span><span class="op">[</span><span class="va">r2</span><span class="op">]</span>, plotType <span class="op">=</span> <span class="st">"line"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ngsReportsIntroduction_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="per-base-sequence-content">Per Base Sequence Content<a class="anchor" aria-label="anchor" href="#per-base-sequence-content"></a>
</h4>
<p>The <code>Per_base_sequence_content</code> module can also be plotted
for an individual report with the layout being identical to that from
FastQC.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSeqContent-methods.html">plotSeqContent</a></span><span class="op">(</span><span class="va">fdl</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="ngsReportsIntroduction_files/figure-html/unnamed-chunk-15-1.png" alt="Individual Per_base_sequence_content plot" width="700"><p class="caption">
Individual Per_base_sequence_content plot
</p>
</div>
<p>These are then combined across multiple files as a heatmap showing a
composite colour for each position. Colours are combined using
<code><a href="https://rdrr.io/r/grDevices/rgb.html" class="external-link">rgb()</a></code> with <code>A</code>,<code>C</code>, <code>G</code>
and <code>T</code> being represented by green, blue, black and red
respectively.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSeqContent-methods.html">plotSeqContent</a></span><span class="op">(</span><span class="va">fdl</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="ngsReportsIntroduction_files/figure-html/unnamed-chunk-16-1.png" alt="Combined Per_base_sequence_content plot" width="700"><p class="caption">
Combined Per_base_sequence_content plot
</p>
</div>
<p><strong>NB</strong> These plots can be very informative setting the
argument <code>usePlotly = TRUE</code>, however they can be slow to
render given the nature of how the package <code>plotly</code> renders
graphics.</p>
<p>Again, supplying multiple files and setting
<code>plotType = "line"</code> will replicate multiple individual plots
from the original FastQC reports. The argument <code>nc</code> is passed
to <code><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap()</a></code> from the package <code>ggplot2</code> to
determine the number of columns in the final plot.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSeqContent-methods.html">plotSeqContent</a></span><span class="op">(</span><span class="va">fdl</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, plotType <span class="op">=</span> <span class="st">"line"</span>, nc <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="ngsReportsIntroduction_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="adapter-content">Adapter Content<a class="anchor" aria-label="anchor" href="#adapter-content"></a>
</h4>
<p>Adapter content as identified by FastQC is also able to be plotted
for an individual file.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotAdapterContent-methods.html">plotAdapterContent</a></span><span class="op">(</span><span class="va">fdl</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> </span></code></pre></div>
<div class="figure">
<img src="ngsReportsIntroduction_files/figure-html/unnamed-chunk-18-1.png" alt="Adapter Content plot for a single FastQC report" width="700"><p class="caption">
Adapter Content plot for a single FastQC report
</p>
</div>
<p>When producing a heatmap across a set of FastQC reports, this will
default to Total Adapter Content, instead of showing the individual
adapter types.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotAdapterContent-methods.html">plotAdapterContent</a></span><span class="op">(</span><span class="va">fdl</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="ngsReportsIntroduction_files/figure-html/unnamed-chunk-19-1.png" alt="Heatmap showing Total Adapter Content by position across a set of FastQC reports" width="700"><p class="caption">
Heatmap showing Total Adapter Content by position across a set of FastQC
reports
</p>
</div>
</div>
<div class="section level4">
<h4 id="sequence-duplication-levels">Sequence Duplication Levels<a class="anchor" aria-label="anchor" href="#sequence-duplication-levels"></a>
</h4>
<p>As with all modules, the Sequence Duplication Levels plot is able to
be replicated for an individual file.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotDupLevels-methods.html">plotDupLevels</a></span><span class="op">(</span><span class="va">fdl</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="ngsReportsIntroduction_files/figure-html/unnamed-chunk-20-1.png" alt="Example Sequence Duplication Levels plot for an individual file." width="700"><p class="caption">
Example Sequence Duplication Levels plot for an individual file.
</p>
</div>
<p>When plotting across multiple FastQC reports, duplication levels are
shown as a heatmap based on each default bin included in the initial
FastQC reports. By default, the plotted values are the “Pre”
de-duplication values. Note that values are converted to percentages
instead of read numbers to ensure comparability across files. In the
plot below, <code>CCGC_R2</code> shows very low duplication levels,
whilst <code>ATTG_R1</code> shows high levels of duplication. The
commonly observed ‘spikes’ around <code>&gt;10</code> are also evident
as the larger red blocks.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotDupLevels-methods.html">plotDupLevels</a></span><span class="op">(</span><span class="va">fdl</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="ngsReportsIntroduction_files/figure-html/unnamed-chunk-21-1.png" alt="Sequence Duplication Levels for multiple files" width="700"><p class="caption">
Sequence Duplication Levels for multiple files
</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="gc-content">GC content<a class="anchor" aria-label="anchor" href="#gc-content"></a>
</h3>
<div class="section level4">
<h4 id="the-class-theoreticalgc">The class TheoreticalGC<a class="anchor" aria-label="anchor" href="#the-class-theoreticalgc"></a>
</h4>
<p>A selection of Theoretical GC content is supplied with the package in
the object <code>gcTheoretical</code>, which has been defined with the
additional <code>S4</code> class <code>TheoreticalGC</code>. GC content
was calculated using scripts obtained from<br><a href="https://github.com/mikelove/fastqcTheoreticalGC" class="external-link uri">https://github.com/mikelove/fastqcTheoreticalGC</a>.
Available genomes and transcriptomes can be obtained using the function
<code><a href="../reference/gcAvail.html">gcAvail()</a></code> on the object <code>gcTheoretical</code> and
specifying the type.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gcAvail.html">gcAvail</a></span><span class="op">(</span><span class="va">gcTheoretical</span>, <span class="st">"Genome"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 24 × 4</span></span></span>
<span><span class="co">##    Name          Group    Source Version   </span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> Alyrata       Plants   JGI    V1.0      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> Amellifera    Animals  UCSC   apiMel2   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> Athaliana     Plants   TAIR   Arapot11  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> Btaurus       Animals  UCSC   bosTau8   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> Celegans      Animals  UCSC   ce11      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> Cfamiliaris   Animals  UCSC   canFam3   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> Dmelanogaster Animals  UCSC   dm6       </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> Drerio        Animals  UCSC   danRer10  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> Ecoli         Bacteria NCBI   2008/08/05</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> Gaculeatus    Other    UCSC   gasAcu1   </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 14 more rows</span></span></span></code></pre>
</div>
<div class="section level4">
<h4 id="inspecting-gc-content">Inspecting GC Content<a class="anchor" aria-label="anchor" href="#inspecting-gc-content"></a>
</h4>
<p>As with all modules, data for an individual file replicates the
default plot from a FastQC report, but with one key difference. This is
that the Theoretical GC content has been provided in the object
<code>gcTheoretical</code> based on 100bp reads. This empirically
determined content is shown as the Theoretical GC content line.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotGcContent-methods.html">plotGcContent</a></span><span class="op">(</span><span class="va">fdl</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, species <span class="op">=</span> <span class="st">"Hsapiens"</span>, gcType <span class="op">=</span> <span class="st">"Transcriptome"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="ngsReportsIntroduction_files/figure-html/unnamed-chunk-23-1.png" alt="Example GC Content plot using the Hsapiens Transcriptome for the theoretical distribution." width="700"><p class="caption">
Example GC Content plot using the Hsapiens Transcriptome for the
theoretical distribution.
</p>
</div>
<p>Again, data is summarised as a heatmap when plotting across multiple
reports, with the default value being the difference between the
observed and the theoretical GC content.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotGcContent-methods.html">plotGcContent</a></span><span class="op">(</span><span class="va">fdl</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="ngsReportsIntroduction_files/figure-html/unnamed-chunk-24-1.png" alt="Example GC content showing the difference between observed and theoretical GC content across multiple files." width="700"><p class="caption">
Example GC content showing the difference between observed and
theoretical GC content across multiple files.
</p>
</div>
<p>Line plots can also be produce an alternative viewpoint, with read
totals displayed as percentages instead of raw counts.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotGcContent-methods.html">plotGcContent</a></span><span class="op">(</span><span class="va">fdl</span>, plotType <span class="op">=</span> <span class="st">"line"</span>,  gcType <span class="op">=</span> <span class="st">"Transcriptome"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="ngsReportsIntroduction_files/figure-html/unnamed-chunk-25-1.png" alt="Example GC content plot represented as a line plot instead of a heatmap." width="700"><p class="caption">
Example GC content plot represented as a line plot instead of a heatmap.
</p>
</div>
<p>Customized theoretical GC content can be generated using input DNA
sequences from a supplied fasta file.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">faFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>    <span class="st">"extdata"</span>, <span class="st">"Athaliana.TAIR10.tRNA.fasta"</span>, </span>
<span>    package <span class="op">=</span> <span class="st">"ngsReports"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotGcContent-methods.html">plotGcContent</a></span><span class="op">(</span><span class="va">fdl</span>, Fastafile <span class="op">=</span> <span class="va">faFile</span>, n <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="overrepresented-sequences">Overrepresented Sequences<a class="anchor" aria-label="anchor" href="#overrepresented-sequences"></a>
</h3>
<p>When inspecting the Overrepresented Sequence module, the top
<code>n</code>can be plotted for an individual file, again broken down
by their possible source, and coloured based on their
<code>WARN/FAIL</code> status.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotOverrep-methods.html">plotOverrep</a></span><span class="op">(</span><span class="va">fdl</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="ngsReportsIntroduction_files/figure-html/unnamed-chunk-27-1.png" width="700"></p>
<p>When applying this across multiple files, instead of identifying
common sequences across a set of libraries, overrepresented sequences
are summarised by their possible source as defined by FastQC.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotOverrep-methods.html">plotOverrep</a></span><span class="op">(</span><span class="va">fdl</span><span class="op">)</span></span></code></pre></div>
<p><img src="ngsReportsIntroduction_files/figure-html/unnamed-chunk-28-1.png" width="700"></p>
<p>In addition to the above, the most abundant <code>n</code>
overrepresented sequences can be exported as a FASTA file for easy
submission to <code>blast</code>.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/overRep2Fasta-methods.html">overRep2Fasta</a></span><span class="op">(</span><span class="va">fdl</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="importing-log-files">Importing Log Files<a class="anchor" aria-label="anchor" href="#importing-log-files"></a>
</h2>
<p>A selection of log files as produced by tools such as
<code>STAR</code>, <code>hisat2</code>, <code>bowtie</code>,
<code>bowtie2</code> and <code>picard duplicationMetrics</code>, can be
easily imported using the <code><a href="../reference/importNgsLogs.html">importNgsLogs()</a></code> function.</p>
<p>Tool can be specified by the user using the argument
<code>type</code>, however if no <code>type</code> is provided we will
attempt to auto-detect from the file’s structure. Note: only a single
log file type can be imported at any time.</p>
<p>The <code><a href="../reference/importNgsLogs.html">importNgsLogs()</a></code> function currently supports log
files from the following tools.</p>
<p>Adapter removal and trimming</p>
<ul>
<li>AdapterRemoval</li>
<li>cutadapt</li>
<li>trimmomatic</li>
</ul>
<p>Mapping and alignment</p>
<ul>
<li>bowtie</li>
<li>bowtie2</li>
<li>hisat2</li>
<li>samtools flagstat</li>
<li>STAR</li>
<li>picard MarkDuplicates</li>
</ul>
<p>Transcript/gene quantification</p>
<ul>
<li>featureCounts</li>
</ul>
<p>Genome assembly</p>
<ul>
<li>BUSCO</li>
<li>Quast</li>
</ul>
<div class="section level3">
<h3 id="adapter-removal-and-trimming">Adapter removal and trimming<a class="anchor" aria-label="anchor" href="#adapter-removal-and-trimming"></a>
</h3>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Sample1.trimmomaticPE.txt"</span><span class="op">)</span></span>
<span><span class="va">trimmomaticLogs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="va">fl</span>, package <span class="op">=</span> <span class="st">"ngsReports"</span><span class="op">)</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importNgsLogs.html">importNgsLogs</a></span><span class="op">(</span><span class="va">trimmomaticLogs</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="st">"Filename"</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">contains</a></span><span class="op">(</span><span class="st">"Surviving"</span><span class="op">)</span>, <span class="st">"Dropped"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/pander/man/pander.html" class="external-link">pander</a></span><span class="op">(</span></span>
<span>        split.tables <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>        style <span class="op">=</span> <span class="st">"rmarkdown"</span>, </span>
<span>        big.mark <span class="op">=</span> <span class="st">","</span>,</span>
<span>        caption <span class="op">=</span> <span class="st">"Select columns as an example of output from trimmomatic."</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<table class="table">
<caption>Select columns as an example of output from
trimmomatic.</caption>
<colgroup>
<col width="27%">
<col width="16%">
<col width="24%">
<col width="24%">
<col width="9%">
</colgroup>
<thead><tr class="header">
<th align="center">Filename</th>
<th align="center">Both_Surviving</th>
<th align="center">Forward_Only_Surviving</th>
<th align="center">Reverse_Only_Surviving</th>
<th align="center">Dropped</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center">Sample1.trimmomaticPE.txt</td>
<td align="center">38243521</td>
<td align="center">2791713</td>
<td align="center">1464192</td>
<td align="center">1663808</td>
</tr></tbody>
</table>
</div>
<div class="section level3">
<h3 id="mapping-and-alignment">Mapping and alignment<a class="anchor" aria-label="anchor" href="#mapping-and-alignment"></a>
</h3>
<div class="section level4">
<h4 id="log-file-import">Log file import<a class="anchor" aria-label="anchor" href="#log-file-import"></a>
</h4>
<p>Bowtie log files can be parsed and imported</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"bowtiePE.txt"</span>, <span class="st">"bowtieSE.txt"</span><span class="op">)</span></span>
<span><span class="va">bowtieLogs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="va">fls</span>, package <span class="op">=</span> <span class="st">"ngsReports"</span><span class="op">)</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importNgsLogs.html">importNgsLogs</a></span><span class="op">(</span><span class="va">bowtieLogs</span>, type <span class="op">=</span> <span class="st">"bowtie"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="st">"Filename"</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">starts_with</a></span><span class="op">(</span><span class="st">"Reads"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/pander/man/pander.html" class="external-link">pander</a></span><span class="op">(</span></span>
<span>        split.tables <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>        style <span class="op">=</span> <span class="st">"rmarkdown"</span>, </span>
<span>        big.mark <span class="op">=</span> <span class="st">","</span>,</span>
<span>        caption <span class="op">=</span> <span class="st">"Select columns as an example of output from bowtie."</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<table class="table">
<caption>Select columns as an example of output from bowtie.</caption>
<colgroup>
<col width="9%">
<col width="11%">
<col width="29%">
<col width="19%">
<col width="29%">
</colgroup>
<thead><tr class="header">
<th align="center">Filename</th>
<th align="center">Reads_Processed</th>
<th align="center">Reads_With_At_Least_One_Reported_Alignment</th>
<th align="center">Reads_That_Failed_To_Align</th>
<th align="center">Reads_With_Alignments_Suppressed_Due_To_-m</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">bowtiePE.txt</td>
<td align="center">42,867,915</td>
<td align="center">21,891,058</td>
<td align="center">20,748,783</td>
<td align="center">228,074</td>
</tr>
<tr class="even">
<td align="center">bowtieSE.txt</td>
<td align="center">440,372</td>
<td align="center">193,962</td>
<td align="center">246,410</td>
<td align="center">NA</td>
</tr>
</tbody>
</table>
<p>STAR log files can be parsed and imported</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">starLog</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"log.final.out"</span>, package <span class="op">=</span> <span class="st">"ngsReports"</span><span class="op">)</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importNgsLogs.html">importNgsLogs</a></span><span class="op">(</span><span class="va">starLog</span>, type <span class="op">=</span> <span class="st">"star"</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<caption>Select columns as output from STAR</caption>
<colgroup>
<col width="19%">
<col width="39%">
<col width="40%">
</colgroup>
<thead><tr class="header">
<th align="center">Filename</th>
<th align="center">Uniquely_Mapped_Reads_Number</th>
<th align="center">Uniquely_Mapped_Reads_Percent</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center">log.final.out</td>
<td align="center">68,471,490</td>
<td align="center">85.84</td>
</tr></tbody>
</table>
<p>The output of the <code>samtools flagstat</code> module can be parsed
and imported</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">flagstatLog</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"flagstat.txt"</span>, package <span class="op">=</span> <span class="st">"ngsReports"</span><span class="op">)</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importNgsLogs.html">importNgsLogs</a></span><span class="op">(</span><span class="va">flagstatLog</span>, type <span class="op">=</span> <span class="st">"flagstat"</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<caption>Flagstat output for a single file</caption>
<colgroup>
<col width="16%">
<col width="14%">
<col width="13%">
<col width="55%">
</colgroup>
<thead><tr class="header">
<th align="center">Filename</th>
<th align="center">QC-passed</th>
<th align="center">QC-failed</th>
<th align="center">flag</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">flagstat.txt</td>
<td align="center">67,405,447</td>
<td align="center">0</td>
<td align="center">in total</td>
</tr>
<tr class="even">
<td align="center">flagstat.txt</td>
<td align="center">5,275,834</td>
<td align="center">0</td>
<td align="center">secondary</td>
</tr>
<tr class="odd">
<td align="center">flagstat.txt</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">supplementary</td>
</tr>
<tr class="even">
<td align="center">flagstat.txt</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">duplicates</td>
</tr>
<tr class="odd">
<td align="center">flagstat.txt</td>
<td align="center">67,405,447</td>
<td align="center">0</td>
<td align="center">mapped</td>
</tr>
<tr class="even">
<td align="center">flagstat.txt</td>
<td align="center">62,129,613</td>
<td align="center">0</td>
<td align="center">paired in sequencing</td>
</tr>
<tr class="odd">
<td align="center">flagstat.txt</td>
<td align="center">31,065,014</td>
<td align="center">0</td>
<td align="center">read1</td>
</tr>
<tr class="even">
<td align="center">flagstat.txt</td>
<td align="center">31,064,599</td>
<td align="center">0</td>
<td align="center">read2</td>
</tr>
<tr class="odd">
<td align="center">flagstat.txt</td>
<td align="center">62,128,918</td>
<td align="center">0</td>
<td align="center">properly paired</td>
</tr>
<tr class="even">
<td align="center">flagstat.txt</td>
<td align="center">62,128,918</td>
<td align="center">0</td>
<td align="center">with itself and mate mapped</td>
</tr>
<tr class="odd">
<td align="center">flagstat.txt</td>
<td align="center">695</td>
<td align="center">0</td>
<td align="center">singletons</td>
</tr>
<tr class="even">
<td align="center">flagstat.txt</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">with mate mapped to a different chr</td>
</tr>
<tr class="odd">
<td align="center">flagstat.txt</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">with mate mapped to a different chr (mapQ&gt;=5)</td>
</tr>
</tbody>
</table>
<p>In addition to the files produced by the above alignment tools, the
output from Duplication Metrics (<code>picard</code>) can also be
imported. This is imported as a list with a <code>tibble</code>
containing the detailed output in the list element <code>$metrics</code>
and the histogram data included as the second elements.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sysDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"ngsReports"</span><span class="op">)</span></span>
<span><span class="va">fl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">sysDir</span>, <span class="st">"Dedup_metrics.txt"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">dupMetrics</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importNgsLogs.html">importNgsLogs</a></span><span class="op">(</span><span class="va">fl</span>, type <span class="op">=</span> <span class="st">"duplicationMetrics"</span>, which <span class="op">=</span> <span class="st">"metrics"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">dupMetrics</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## tibble [1 × 10] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">##  $ Library                       : chr "~/myExpt/2_alignedData/bams/Sample1.bam"</span></span>
<span><span class="co">##  $ Unpaired Reads Examined       : int 93</span></span>
<span><span class="co">##  $ Read Pairs Examined           : int 43490914</span></span>
<span><span class="co">##  $ Secondary Or Supplementary Rds: int 5366030</span></span>
<span><span class="co">##  $ Unmapped Reads                : int 0</span></span>
<span><span class="co">##  $ Unpaired Read Duplicates      : int 79</span></span>
<span><span class="co">##  $ Read Pair Duplicates          : int 11494386</span></span>
<span><span class="co">##  $ Read Pair Optical Duplicates  : int 642648</span></span>
<span><span class="co">##  $ Percent Duplication           : num 0.264</span></span>
<span><span class="co">##  $ Estimated Library Size        : int 69609522</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="plot-log-file-imports">Plot log file imports<a class="anchor" aria-label="anchor" href="#plot-log-file-imports"></a>
</h4>
<p>Summaries of log files from select mapping and alignment tools can be
plot using the function <code>plotAlignemntSummary()</code>.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotAlignmentSummary.html">plotAlignmentSummary</a></span><span class="op">(</span><span class="va">bowtieLogs</span>, type <span class="op">=</span> <span class="st">"bowtie"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="ngsReportsIntroduction_files/figure-html/unnamed-chunk-39-1.png" alt="Example Bowtie logs for PE and SE sequencing" width="700"><p class="caption">
Example Bowtie logs for PE and SE sequencing
</p>
</div>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotAlignmentSummary.html">plotAlignmentSummary</a></span><span class="op">(</span><span class="va">starLog</span>, type <span class="op">=</span> <span class="st">"star"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="ngsReportsIntroduction_files/figure-html/unnamed-chunk-40-1.png" alt="Example STAR aligner logs" width="700"><p class="caption">
Example STAR aligner logs
</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="genome-assembly">Genome assembly<a class="anchor" aria-label="anchor" href="#genome-assembly"></a>
</h3>
<p>Assembly ‘completeness’ and summary statistic information from the
tools BUSCO and quast can also be plot using the function
<code><a href="../reference/plotAssemblyStats.html">plotAssemblyStats()</a></code></p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">buscoLog</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"short_summary_Dmelanogaster_Busco.txt"</span>, package <span class="op">=</span> <span class="st">"ngsReports"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotAssemblyStats.html">plotAssemblyStats</a></span><span class="op">(</span><span class="va">buscoLog</span>, type <span class="op">=</span> <span class="st">"busco"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="ngsReportsIntroduction_files/figure-html/unnamed-chunk-41-1.png" alt="Example plot after running BUSCO v3 on the Drosophila melanogaster reference genome" width="700"><p class="caption">
Example plot after running BUSCO v3 on the Drosophila melanogaster
reference genome
</p>
</div>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"quast1.tsv"</span>, <span class="st">"quast2.tsv"</span><span class="op">)</span></span>
<span><span class="va">quastLog</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="va">fls</span>, package <span class="op">=</span> <span class="st">"ngsReports"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotAssemblyStats.html">plotAssemblyStats</a></span><span class="op">(</span><span class="va">quastLog</span>, type <span class="op">=</span> <span class="st">"quast"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="ngsReportsIntroduction_files/figure-html/unnamed-chunk-42-1.png" alt="Example plot after running quast on two shortread assemblies" width="700"><p class="caption">
Example plot after running quast on two shortread assemblies
</p>
</div>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotAssemblyStats.html">plotAssemblyStats</a></span><span class="op">(</span><span class="va">quastLog</span>, type <span class="op">=</span> <span class="st">"quast"</span>, plotType <span class="op">=</span> <span class="st">"paracoord"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="ngsReportsIntroduction_files/figure-html/unnamed-chunk-43-1.png" alt="Example parallel coordinate plot after running quast on two shortread assemblies" width="700"><p class="caption">
Example parallel coordinate plot after running quast on two shortread
assemblies
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">## R Under development (unstable) (2024-02-14 r85901)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] pander_0.6.5        dplyr_1.1.4         ngsReports_2.5.3   </span></span>
<span><span class="co">## [4] tibble_3.2.1        patchwork_1.2.0     ggplot2_3.4.4      </span></span>
<span><span class="co">## [7] BiocGenerics_0.49.1 BiocStyle_2.31.0   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] gtable_0.3.4            xfun_0.42               bslib_0.6.1            </span></span>
<span><span class="co">##  [4] htmlwidgets_1.6.4       lattice_0.22-5          vctrs_0.6.5            </span></span>
<span><span class="co">##  [7] tools_4.4.0             bitops_1.0-7            generics_0.1.3         </span></span>
<span><span class="co">## [10] stats4_4.4.0            fansi_1.0.6             highr_0.10             </span></span>
<span><span class="co">## [13] pkgconfig_2.0.3         data.table_1.15.0       RColorBrewer_1.1-3     </span></span>
<span><span class="co">## [16] desc_1.4.3              S4Vectors_0.41.3        lifecycle_1.0.4        </span></span>
<span><span class="co">## [19] GenomeInfoDbData_1.2.11 farver_2.1.1            compiler_4.4.0         </span></span>
<span><span class="co">## [22] stringr_1.5.1           textshaping_0.3.7       Biostrings_2.71.2      </span></span>
<span><span class="co">## [25] munsell_0.5.0           GenomeInfoDb_1.39.6     htmltools_0.5.7        </span></span>
<span><span class="co">## [28] sass_0.4.8              RCurl_1.98-1.14         yaml_2.3.8             </span></span>
<span><span class="co">## [31] lazyeval_0.2.2          plotly_4.10.4           tidyr_1.3.1            </span></span>
<span><span class="co">## [34] pillar_1.9.0            pkgdown_2.0.7           crayon_1.5.2           </span></span>
<span><span class="co">## [37] jquerylib_0.1.4         MASS_7.3-60.2           cachem_1.0.8           </span></span>
<span><span class="co">## [40] tidyselect_1.2.0        digest_0.6.34           stringi_1.8.3          </span></span>
<span><span class="co">## [43] reshape2_1.4.4          purrr_1.0.2             bookdown_0.37          </span></span>
<span><span class="co">## [46] labeling_0.4.3          forcats_1.0.0           fastmap_1.1.1          </span></span>
<span><span class="co">## [49] grid_4.4.0              colorspace_2.1-0        cli_3.6.2              </span></span>
<span><span class="co">## [52] magrittr_2.0.3          utf8_1.2.4              withr_3.0.0            </span></span>
<span><span class="co">## [55] scales_1.3.0            lubridate_1.9.3         timechange_0.3.0       </span></span>
<span><span class="co">## [58] rmarkdown_2.25          XVector_0.43.1          httr_1.4.7             </span></span>
<span><span class="co">## [61] zoo_1.8-12              ragg_1.2.7              memoise_2.0.1          </span></span>
<span><span class="co">## [64] evaluate_0.23           knitr_1.45              IRanges_2.37.1         </span></span>
<span><span class="co">## [67] viridisLite_0.4.2       rlang_1.1.3             ggdendro_0.1.23        </span></span>
<span><span class="co">## [70] Rcpp_1.0.12             glue_1.7.0              BiocManager_1.30.22    </span></span>
<span><span class="co">## [73] jsonlite_1.8.8          plyr_1.8.9              R6_2.5.1               </span></span>
<span><span class="co">## [76] systemfonts_1.0.5       fs_1.6.3                zlibbioc_1.49.0</span></span></code></pre>
</div>
<div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" class="external-link uri">https://www.bioinformatics.babraham.ac.uk/projects/fastqc/</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Stevie Pederson, <a href="https://github.com/CMWbio" class="external-link">Christopher Ward</a>, <a href="https://github.com/tothuhien" class="external-link">Thu-Hien To</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
